src/Test/L0/Worker/Build/BuildJobExtensionL0.cs

```diff
@@ -0,0 +1,236 @@
+        private Pipelines.WorkspaceOptions _workspaceOptions;
+        private char directorySeparator = Path.DirectorySeparatorChar;
+
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Worker")]
+        public void CheckSingleRepoWithoutPathInput()
+        {
+            using (TestHostContext tc = Setup(createWorkDirectory: false, checkOutConfig: CheckoutConfigType.SingleCheckoutDefaultPath))
+            {
+                buildJobExtension.InitializeJobExtension(_ec.Object, steps, _workspaceOptions);
+                var repoLocalPath = _ec.Object.Variables.Get(Constants.Variables.Build.RepoLocalPath);
+                Assert.NotNull(repoLocalPath);
+                Assert.Equal(Path.Combine(stubWorkFolder, $"1{directorySeparator}s"), repoLocalPath);
+            }
+        }
+
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Worker")]
+        public void CheckSingleRepoWithCustomPaths()
+        {
+            using (TestHostContext tc = Setup(createWorkDirectory: false, checkOutConfig: CheckoutConfigType.SingleCheckoutCustomPath, pathToSelfRepo: "s/CustomApplicationFolder"))
+            {
+                buildJobExtension.InitializeJobExtension(_ec.Object, steps, _workspaceOptions);
+                var repoLocalPath = _ec.Object.Variables.Get(Constants.Variables.Build.RepoLocalPath);
+                Assert.NotNull(repoLocalPath);
+                Assert.Equal(Path.Combine(stubWorkFolder, $"1{directorySeparator}s"), repoLocalPath);
+            }
+        }
+
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Worker")]
+        public void CheckMultiRepoWithoutPathInput()
+        {
+            using (TestHostContext tc = Setup(createWorkDirectory: false, checkOutConfig: CheckoutConfigType.MultiCheckoutDefaultPath))
+            {
+                buildJobExtension.InitializeJobExtension(_ec.Object, steps, _workspaceOptions);
+                var repoLocalPath = _ec.Object.Variables.Get(Constants.Variables.Build.RepoLocalPath);
+                Assert.NotNull(repoLocalPath);
+                Assert.Equal(Path.Combine(stubWorkFolder, $"1{directorySeparator}s"), repoLocalPath);
+            }
+        }
+
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Worker")]
+        public void CheckMultiRepoWithPathInputToCustomPath()
+        {
+            using (TestHostContext tc = Setup(createWorkDirectory: false, checkOutConfig: CheckoutConfigType.MultiCheckoutCustomPath, pathToSelfRepo: "s/CustomApplicationFolder"))
+            {
+                buildJobExtension.InitializeJobExtension(_ec.Object, steps, _workspaceOptions);
+                var repoLocalPath = _ec.Object.Variables.Get(Constants.Variables.Build.RepoLocalPath);
+                Assert.NotNull(repoLocalPath);
+                Assert.Equal(Path.Combine(stubWorkFolder, $"1{directorySeparator}s{directorySeparator}App"), repoLocalPath);
+            }
+        }
```

Comment:
@EzzhevNikita Are these unit tests meant to specify the previous behavior and demonstrate it hasn't changed? I copied just `BuildJobExtensionL0.cs` to `master` and ran it and got one failure. Can you explain that? I was expecting all of these to pass on `master` without any of the other code changes since the intent is to not change the behavior of the variable, but maybe I'm missing something.

```
[xUnit.net 00:00:04.7828244]     Microsoft.VisualStudio.Services.Agent.Tests.Worker.Build.BuildJobExtensionL0.CheckMultiRepoWithPathInputToCustomPath [FAIL]
    X Microsoft.VisualStudio.Services.Agent.Tests.Worker.Build.BuildJobExtensionL0.CheckMultiRepoWithPathInputToCustomPath [216ms]
EXEC : error Message:  [C:\Users\maroghel\src\azure-pipelines-agent\src\dir.proj]
     Assert.Equal() Failure
                                   ↓ (pos 82)
  Expected: ···c9470121d8f\Work\1\s\App
  Actual:   ···c9470121d8f\Work\1\s
                                   ↑ (pos 82)
```

Reply:
@mjroghelia Let me a little bit describe the issue itself, that will help to understand the idea of these tests.
The problem with the changed value of the variable is the result of the behavior of three modules: BuildDirectoryManager, BuildJobExtension, and PluginInternalCommandExtension.

BuildJobExtension previously set the default value for Build.RepoLocalPath to the SourcesDirectory from tracking config.
PluginInternalCommandExtension can update this variable in case the checkout task moved the repository, in other words, the current location of the repository is not the same as the checkout task computed based on the path from the input.

* Before we fixed the bug in BuildDirectoryManager, PluginInternalCommandExtension performs updating of the Build.RepoLocalPath variable in multi checkout jobs on every checkout of repository for which was set custom path. It allows the variable to point to the root of the self repository.
* When we fixed BuildDirectoryManager and allow it to rely on tracking config for a particular repository, moving repositories began to occur only in cases when its path really changes while reusing (which is not happens in most cases). Because of it, we've started seeing that for the Jobs where there are several checkouts with custom paths Build.RepoLocalPath variable started to point to the sources directory instead of the root of the self repository.

I think that tests that fully check the behavior of this variable must cover the logic of all three components, which is why I wrote in RCA that canary tests would be the best option for this problem.

Tests that I've added are checking that we are setting the path for the repositories from tracking config only in case it is a job with multiple checkouts and for the repository was set input contains the non-default path. And they shouldn't pass in master.

Reply:
@EzzhevNikita Thanks for the elaboration!